src/client_conn.c

```diff
@@ -112,43 +109,47 @@ void tetris_connect(NetClient *net_client, char *host, int port) {
 	init_sockaddr(&servername, host, port);
 	if (0 > connect(sock_fd, (struct sockaddr *)&servername,
 	                sizeof(servername))) {
-		perror("connect (client)");
-		exit(EXIT_FAILURE);
+		return EXIT_FAILURE;
 	}
 
 	net_client->fd = sock_fd;
+
+	return EXIT_SUCCESS;
 }
 
 /**
  * register
  */
-void tetris_register(NetClient *net_client, char *username) {
-	char message[128];
-	sprintf(message, "%c%s", MSG_TYPE_REGISTER, username);
-	tetris_send_message(net_client, message);
+NetRequest *tetris_register(NetClient *net_client, char *username) {
+	char message[16];
+	strncpy(message, username, 16);
+	return ttetris_net_request(net_client, message, 16, MSG_TYPE_REGISTER);
 }
 
 /**
  * select opponent by username
  */
-void tetris_opponent(NetClient *net_client, char *username) {
-	char message[128];
-	sprintf(message, "%c%s", MSG_TYPE_OPPONENT, username);
-	tetris_send_message(net_client, message);
+void tetris_opponent(NetClient *net_client, StringArray *usernames) {
+	Blob *message = string_array_serialize(usernames);
+	ttetris_net_request(net_client, message->bytes, message->length,
+	                    MSG_TYPE_OPPONENT);
+	free(message->bytes);
+	free(message);
+}
+
+void tetris_tell_server_to_start(NetClient *net_client) {
+	message_nbytes(net_client->fd, NULL, 0, 0, MSG_TYPE_START_GAME);
 }
 
 void tetris_disconnect(NetClient *net_client) {
 	close(net_client->fd);
 	net_client->is_listen_thread_started = 0;
 }
 
-int read_game_view_data(char **cursor, struct game_view_data *view) {
-	char *buffer = *cursor;
-	// move the pointer past the message identifier
-	++buffer;
+static int read_game_view_data(char *buffer, struct game_view_data *view) {
 	// get the player associated with the board
 	char *board_name = buffer;
-	int name_length = strlen(board_name);
+	unsigned long name_length = strnlen(board_name, PLAYER_NAME_MAX_CHARS);
```

Comment:
NIT: Agree with unsigned usage. But I hope that name length doesn't exceed the size of an int. Is there a reason to use a long here?

Reply:
That's fair; I used a long because `strnlen` returns a `size_t`, which on most systems equates to an unsigned long afaik.

Of course, in practice `strnlen` will return a number smaller than `PLAYER_NAME_MAX_CHARS`, which is currently equal to 15, so I guess we could even use an `unsigned char`.

I will use `unsigned int` for now, per your suggestion. :smile:

Reply:
That is interesting, didn't look at the output of strnlen!